src/Workspaces/Core/Portable/NamingStyles/NamingStyle.cs

```diff
@@ -0,0 +1,368 @@
+            => CheckAllWords(
+                name, nameSpan, s_wordIsAllLowerCase,
+                WorkspacesResources.These_words_cannot_contain_lower_case_characters_colon_0, out reason);
+
+        private bool CheckFirstAndRestWords(
+            string name, TextSpan nameSpan,
+            Func<string, TextSpan, bool> firstWordCheck,
+            Func<string, TextSpan, bool> restWordCheck,
+            string firstResourceId,
+            string restResourceId,
+            out string reason)
+        {
+            reason = null;
+            var violations = ArrayBuilder<string>.GetInstance();
+
+            var first = true;
+
+            foreach (var wordSpan in GetWordSpans(name, nameSpan))
+            {
+                if (first)
+                {
+                    if (!firstWordCheck(name, wordSpan))
+                    {
+                        reason = string.Format(firstResourceId, Substring(name, wordSpan));
+                    }
+                }
+                else
+                {
+                    if (!restWordCheck(name, wordSpan))
+                    {
+                        violations.Add(Substring(name, wordSpan));
+                    }
+                }
+
+                first = false;
+            }
+
+            if (violations.Count > 0)
+            {
+                var restString = string.Format(restResourceId, string.Join(", ", violations));
+                reason = reason == null
+                    ? restString
+                    : reason + Environment.NewLine + restString;
+            }
+
+            violations.Free();
+            return reason == null;
+        }
+
+        private bool CheckCamelCase(string name, TextSpan nameSpan, out string reason)
+            => CheckFirstAndRestWords(
+                name, nameSpan, s_firstCharIsLowerCase, s_firstCharIsUpperCase,
+                WorkspacesResources.The_first_word_0_must_begin_with_a_lower_case_character,
+                WorkspacesResources.These_non_leading_words_must_begin_with_an_upper_case_letter_colon_0,
+                out reason);
+
+        private bool CheckFirstUpper(string name, TextSpan nameSpan, out string reason)
+            => CheckFirstAndRestWords(
+                name, nameSpan, s_firstCharIsUpperCase, s_firstCharIsLowerCase,
+                WorkspacesResources.The_first_word_0_must_begin_with_an_upper_case_character,
+                WorkspacesResources.These_non_leading_words_must_begin_with_a_lowercase_letter_colon_0,
+                out reason);
+
+        private string CreateCompliantNameDirectly(string name)
+        {
+            var addPrefix = !name.StartsWith(Prefix);
+            var addSuffix = !name.EndsWith(Suffix);
+
+            name = addPrefix ? (Prefix + name) : name;
+            name = addSuffix ? (name + Suffix) : name;
+
+            return FinishFixingName(name);
+        }
+
+        public IEnumerable<string> MakeCompliant(string name)
```

Comment:
I don't want to touch any of hte 'fixing' code in this PR.  only the analysis code.  That's the code that's going to run non-stop on every symbol after every edit.